src/jdk.naming.dns/share/classes/com/sun/jndi/dns/DnsClient.java

```diff
@@ -402,43 +428,72 @@ private byte[] doUdpQuery(Packet pkt, InetAddress server,
                                      int port, int retry, int xid)
             throws IOException, NamingException {
 
-        int minTimeout = 50; // msec after which there are no retries.
-
-        synchronized (udpSocketLock) {
-            try (DatagramSocket udpSocket = getDatagramSocket()) {
-                DatagramPacket opkt = new DatagramPacket(
-                        pkt.getData(), pkt.length(), server, port);
-                DatagramPacket ipkt = new DatagramPacket(new byte[8000], 8000);
+        udpChannelLock.lock();
+        try {
+            try (DatagramChannel udpChannel = getDatagramChannel()) {
+                ByteBuffer opkt = ByteBuffer.wrap(pkt.getData(), 0, pkt.length());
+                byte[] data = new byte[8000];
+                ByteBuffer ipkt = ByteBuffer.wrap(data);
                 // Packets may only be sent to or received from this server address
-                udpSocket.connect(server, port);
+                InetSocketAddress target = new InetSocketAddress(server, port);
+                udpChannel.connect(target);
                 int pktTimeout = (timeout * (1 << retry));
-                udpSocket.send(opkt);
+                udpChannel.send(opkt, target);
 
                 // timeout remaining after successive 'receive()'
                 int timeoutLeft = pktTimeout;
                 int cnt = 0;
+                boolean gotData = false;
                 do {
+                    // prepare for retry
+                    if (gotData) {
+                        Arrays.fill(data, 0, ipkt.position(), (byte) 0);
+                        ipkt.clear();
+                    }
                     if (debug) {
                         cnt++;
                         dprint("Trying RECEIVE(" +
                                 cnt + ") retry(" + (retry + 1) +
                                 ") for:" + xid + "    sock-timeout:" +
                                 timeoutLeft + " ms.");
                     }
-                    udpSocket.setSoTimeout(timeoutLeft);
                     long start = System.currentTimeMillis();
-                    udpSocket.receive(ipkt);
+                    gotData = blockingReceive(udpChannel, ipkt, timeoutLeft);
                     long end = System.currentTimeMillis();
-
-                    byte[] data = ipkt.getData();
-                    if (isMatchResponse(data, xid)) {
+                    assert gotData || ipkt.position() == 0;
+                    if (gotData && isMatchResponse(data, xid)) {
                         return data;
                     }
                     timeoutLeft = pktTimeout - ((int) (end - start));
-                } while (timeoutLeft > minTimeout);
-                return null; // no matching packet received within the timeout
+                } while (timeoutLeft > MIN_TIMEOUT);
+                // no matching packet received within the timeout
+                throw new SocketTimeoutException();
```

Comment:
It appears to me that, before the change in this PR, we used to return `null` from this method if there is a timeout. The calling code (the method `query`) would then interpret this `null` return in a couple of different ways. One of them being, skipping this server and querying any other server(s) that were known to the client instance. Now, with this change where we throw this `SocketTimeoutException`, that part of the code would behave differently from what I can see. Is this intentional to throw an exception instead of returning `null`?

Reply:
Thanks for spotting that, Jai.
You're correct that in its current version the fix changed an old logic of `doUdpQuery`/`query` methods:
Before this change the method was returning `null` not when a receive is timed out but when an unmatched packet is received. Socket timeout exceptions thrown by `DatagramSocket.receive` were caught in `query` method.
After the proposed change the `doUdpQuery` method is throwing `SocketTimeoutException` for both cases (timeout and unmatched packets) - that needs to be changed to comply with old logic. Will address it in an upcoming changeset.

Reply:
This is addressed now - right?

Reply:
Yes, already addressed in 55dd0a4